File: week11/338-counting-bits/countBits.cpp
class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> result(num + 1, 0);
        int highBit = 0;
        for (int i = 1; i <= num; ++i) {
            if ((i & -i) == i) {
                highBit = i;
                result[i] = 1;
            } else result[i] = result[i & ~highBit] + 1;
        }
        return result;
    }
};

File: week11/338-counting-bits/countBits.scala
/** It works perfectly on my machine but not on leetcode. But I believe it is right.
* So here is another C++ solution.
*/
object Solution {
  def countBits(num: Int): Array[Int] = {
    val result = collection.mutable.Seq.fill(num + 1)(0)
    1.to(num).foreach { i =>
      result.update(i, result(i & (i - 1)) + 1)
    }
    result.toArray
  }
}